### PR TITLE
[5.0] Casting Nullable Fields

### PR DESCRIPTION
**Pull request by rwillians**
Perhaps casting types should support nullable fields in order not to change the expected output value.
## Issue
- Casting nullable fields to `integer`, `real`, `float` or `double` changes the output value to 0 (zero);
- Casting nullable fields to `string`changes the output value to "" (empty).
## Scenario

Consider you have the model below that is a representation of a mobile phone, where the `null` "carrier_id" represents there is a job pending to find the phone's carrier.

``` json
{
    "id": 1,
    "carrier_id": null,
    "country_id": 55,
    "area": 11,
    "number": 987654321
}
```

Using the following casts:

``` php
protected $casts = [
    'id' => 'integer',
    'carrier_id' => 'integer',
    'country_id' => 'integer',
    'area' => 'integer',
    'number' => 'integer'
];
```

When you output your model (without the suggested changes) you will get:

``` json
{
    "id": 1,
    "carrier_id": 0,
    "country_id": 55,
    "area": 11,
    "number": 987654321
}
```

With the suggested changes you will get the correct expected output:

``` json
{
    "id": 1,
    "carrier_id": null,
    "country_id": 55,
    "area": 11,
    "number": 987654321
}
```

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
Interesting. Could you add some tests please?

**Comment by Arrilot** [comment]
I have some points here:
1) The change should be done for all types, not only `int`, `float` and `string`.
2) There are different points of view on this issue - Personally, I don't like this change because we do not describe our schema in `$casts` property. 
We describe how we want to cast our data instead. 
And if we __explicitly asked__ to cast data to `int` why should we get `null`?
3) I can also propose this making casting behaviour configurable per model and/or globally. Not sure if it's a good idea or not

**Comment by barryvdh** [comment]
Or something like this to describe nullable?

```'carrier_id' => 'integer|null',```

**Comment by Anahkiasen** [comment]
@barryvdh 's solution would seem more inline to me with how Laravel usually declares rules (like for validation) etc

**Comment by Arrilot** [comment]
@barryvdh Yup, even better :+1:

**Comment by barryvdh** [comment]
FYI, this is how it was done in jsonmapper: https://github.com/netresearch/jsonmapper/commit/c83043716f5be254a2a46d4816ade2ccc313589c
Not sure if there are more efficient ways.

**Comment by rwillians** [comment]
My point was not mutate the value, only represent it as it is on database.
If an integer fields should not support null as @Arrilot mentioned, then the field should not be nullable on the database in the first place.
The situation I represented covers the scenario when you are reading from the database and needs to make sure the values are represented properly.
If in some situations you need to mutate the value then a mutator should be created.

**Comment by rwillians** [comment]
@GrahamCampbell, I've added some tests.

**Comment by rwillians** [comment]
@barryvdh using ```integer|null``` seams to be an alternative. In my point of view it's not necessary because a believe you should restrict fields to not accept ```null``` on database.
As a said before, my idea was only to represent the correct value the model got from the database.

File: tests/Database/DatabaseEloquentCastingSupportsNullable.php [hunk 1]
```diff
@@ -0,0 +1,114 @@
+<?php
+
+use Illuminate\Database\Eloquent\Model;
+
+class DatabaseEloquentCastingSupportsNullable extends PHPUnit_Framework_TestCase {
+
+	public function testCastingStringToInteger()
+	{
+		$model = new EloquentModelCastingStub();
+        $model->setAttribute('string_to_integer_field', '1');
```
Comment (GrahamCampbell) [inline tests/Database/DatabaseEloquentCastingSupportsNullable.php]: tabs for indentation please

File: tests/Database/DatabaseEloquentCastingSupportsNullable.php [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    public function testCastingStringToFloat()
+    {
+        $model = new EloquentModelCastingStub();
+        $model->setAttribute('string_to_float_field', '1.0');
+
+        $this->assert(true, is_float($model->getAttribute('string_to_float_field')));
+    }
+
```
Comment (GrahamCampbell) [inline tests/Database/DatabaseEloquentCastingSupportsNullable.php]: You need 2 new lines between test methods please, not 1.

**Comment by rwillians** [comment]
@GrahamCampbell done

**Comment by Arrilot** [comment]
i'm afraid of being annoying, but ... 
> My point was not mutate the value, only represent it as it is on database. If in some situations you need to mutate the value then a mutator should be created.

@rwillians Your point of view diverges from the official documentation http://laravel.com/docs/5.0/eloquent#attribute-casting

> 1) The change should be done for all types, not only int, float and string.

This is still valid.
```php 
protected $casts = ['some_field_that_is_null' => 'string']; => null
```
```php 
protected $casts = ['some_field_that_is_null' => 'array']; => []
```
This is really inconsistent

And, btw this whole implementation is a decent BC break and should probably go to 5.1 branch
@barryvdh 's proposal is BC friendly

cheers

**Comment by rwillians** [comment]
@Arrilot, I sincerely don't see how this diverges from the official cast documentation (or at least its purpose). I might be wrong but in my understanding "cast" should help you change a value's type not the value itself, but that's what is happening (please read the scenario described in the proposal).

You made me thinking and in my conclusion you're right about being inconsistent handling ```null```s only for ```integer```, ```float``` and ```string```. I will implement my proposal to handle it for ```boolean``` as well (not necessary handle for array since the json_encode/json_decode already does that).

My point (again) is: converting a ```null``` into a ```0``` or an ```""``` (empty) is changing the value and not the value's type and doesn't represents the real value stored on the database. That's what my proposal is fixing.

I'm sorry my ignorance but I don't see how this breaks BC.

**Comment by JosephSilber** [comment]
I think if do it this way, then we should return `null` as is for everything: check for `null` before the switch statement, and if it's `null` return it.

**Comment by Arrilot** [comment]
@rwillians I'm glad I helped you a little
> I'm sorry my ignorance but I don't see how this breaks BC

Some ppl (like me) may think that the current behavior is correct and if they for some reason have strict checks in their code e.g. 
```php 
if ($user_id !== 0) {...}
``` 
their code will be broken after composer update.
Not the biggest problem in the world ofc, but should be taken in consideration imho

**Comment by rwillians** [comment]
@Arrilot I see your point, thank you.

I just think you meant: 
```php
if ($user_id != 0) {...}
```
Right?

Because if you are actually expecting integer 0 (zero) it would be your logic error not Laravel's, therefore this PR wouldn't break BC. And if you meant ```!=``` it would still work with ```null```.

What do you think?

**Comment by rwillians** [comment]
@JosephSilber agreed.

**Comment by JosephSilber** [comment]
@rwillians why would `!==` be an error in user-code logic?

If you know you'll be dealing with an integer, doing a strict type check seems quite reasonable.

**Comment by rwillians** [comment]
@JosephSilber because the scenario this PR is fixing if for a field that can be nullable.
If @Arrilot meant ```!==``` in my point of view it is his logic fault not handling the null situation.
If he meant ```!=``` there will be no BC break and will work perfectly.

**Comment by JosephSilber** [comment]
For the record, I do like this solution more than @barryvdh's. But being that this is a BC breaking change, it should go to 5.1.

This is actually what I've been doing in my own base model class:

```php
protected function castAttribute($key, $value)
{
    if (is_null($value)) return null;

    return parent::castAttribute($key, $value);
}
```

**Comment by JosephSilber** [comment]
This looks good now, but...

1. There's no need for this completely new test file. You can just add the `null` tests to [this](https://github.com/laravel/framework/blob/5.0/tests/Database/DatabaseEloquentModelTest.php#L1109).
2. Your whitespace is again wrong. There should be no spaces inside the parentheses of the conditional.
3. Squash those 11 commits down into 1.
4. Close this PR and  open a new one against master.

Good luck!

**Comment by rwillians** [comment]
@mattwells' solution #7531 (as well suggested before by @barryvdh) seems to be a nicer solution since it would not break BC. Therefore, I'm closing this PR.

Thanks for the help!

**Comment by JosephSilber** [comment]
@taylorotwell while I love this feature (and currently use it myself), this is a BC breaking feature that I think should go to 5.1

**Comment by taylorotwell** [comment]
It's a bug fix.

**Pull request merged**